Add Support of Types for GivesBuildableArea

MustaphaTR opened the pull request:
Fixes #10550.

Also moved GivesBuildableArea to its own file from `Building.cs`

~~For testcase i made D2K engineer allow placement of walls around him normally and turrets when owner has palace.~~

MustaphaTR committed:
Add Support of Types for GivesBuildableArea

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -204,12 +203,13 @@ public virtual bool IsCloseEnoughToBase(World world, Player p, string buildingNa
 							&& (a.Owner == p || (allyBuildEnabled && a.Owner.Stances[p] == Stance.Ally))
-							&& a.Info.HasTraitInfo<GivesBuildableAreaInfo>());
+							&& a.Info.HasTraitInfo<GivesBuildableAreaInfo>()
+							&& a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>().ValidTypes.Contains(AdjacentType));
```
You should change the `TraitInfoOrDefault` here and below to only `TraitInfo`, as the `HasTraitInfo` checks already guarantee that the trait is there.

pchote replied:
It would be better if this was pulled out to a `Func<Actor, bool>` and used `TraitInfoOrDefault` and a null check instead of `HasTraitInfo`.

MustaphaTR commented:
NTS: Needs rebase.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/GivesBuildableArea.cs:
```diff
@@ -0,0 +1,29 @@
+		public object Create(ActorInitializer init) { return new GivesBuildableArea(this); }
+	}
+
+	public class GivesBuildableArea {
```
The brace should be on its own line.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -177,6 +176,9 @@ public Actor FindBaseProvider(World world, Player p, CPos topLeft)
 		public virtual bool IsCloseEnoughToBase(World world, Player p, string buildingName, CPos topLeft)
 		{
+			Func<Actor, bool> hasAdjanentType = a => a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>().ValidTypes.Contains(AdjacentType);
+			Func<Actor, bool> givesBuildableAreaNotNull = a => a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>() != null;
```
I think @pchote had one combined function in mind, like:
```
Func<Actor, bool> hasAdjanentType = a =>
{
	var givesBuildableArea = a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>();
	if (givesBuildableArea == null)
		return false;

	return givesBuildableArea.ValidTypes.Count == 0 || givesBuildableArea.ValidType.Contains(AdjacentType);
};
```
(Note: I also added a `Count == 0` to ensure that `Leave empty to accept all types.` works. Did that work before without that check?)

MustaphaTR commented:
Updated.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -173,6 +172,15 @@ public Actor FindBaseProvider(World world, Player p, CPos topLeft)
 
 		public virtual bool IsCloseEnoughToBase(World world, Player p, string buildingName, CPos topLeft)
 		{
+			Func<Actor, bool> hasAdjanentType = a =>
```
Typo here.

RoosterDragon replied:
Also you should move this to be defied closer to the loop where it's used. At the moment if the two following conditions bail out this goes unused.

pchote replied:
Would it be worth pulling this out into a proper method to avoid unnecessary object creation?

RoosterDragon replied:
Yes.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -40,6 +36,9 @@ public class BuildingInfo : ITraitInfo, IOccupySpaceInfo, IPlaceBuildingDecorati
 		public readonly int Adjacent = 2;
 
+		[Desc("Type of influence, defined by the `GivesBuildableArea` traits of the nearby actors, this can be build on.")]
+		public readonly string AdjacentType = null;
```
Can you please make this a `public readonly HashSet<string> AreaTypes = new HashSet<string>();`?

Two things here:
* The *Type name should match the trait name.
* Supporting multiple here gives more flexibility.  The only change later is to switch the `.Contains(AdjacentType)` with `.Overlaps(AreaTypes)`.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -188,6 +187,15 @@ public virtual bool IsCloseEnoughToBase(World world, Player p, string buildingNa
 
+			Func<Actor, bool> hasAdjacentType = a =>
+			{
+				var givesBuildableArea = a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>();
```
Please use `.TraitInterfaces` and then loop over the results.  There's no reason not to support multiple traits on a single actor.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/GivesBuildableArea.cs:
```diff
@@ -0,0 +1,30 @@
+	public class GivesBuildableAreaInfo : ITraitInfo
+	{
+		[Desc("Accepted `AdjacentTypes` types. Leave empty to accept all types.")]
+		public readonly HashSet<string> ValidTypes = new HashSet<string>();
```
Nit: the convention from the other traits is to call this just `Types`.

pchote reviewed (commented):
Sorry for the long delay in reviewing this.  Looks good overall, but a couple of requests to make it a bit more general:

MustaphaTR commented:
Updated and Rebased.

Testcase now gives engineers ability to build turrets around them when owner has Palace to test that conditions on `GivesBuildableArea:` and multipile `GivesBuildableArea:` traits are working.

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -382,4 +382,9 @@ public interface ITargetableCells
 		IEnumerable<Pair<CPos, SubCell>> TargetableCells();
 	}
+
+	public interface IGiveBuildableArea
```
This interface doesn't give us anything, so please just reference `GivesBuildableArea` directly from `BuildingInfo`.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -188,6 +187,25 @@ public virtual bool IsCloseEnoughToBase(World world, Player p, string buildingNa
 			var bi = world.WorldActor.Trait<BuildingInfluence>();
 			var allyBuildEnabled = world.WorldActor.Trait<MapBuildRadius>().AllyBuildRadiusEnabled;
 
+			Func<Actor, bool> hasAreaTypes = a =>
```
Please instead extract a function
```csharp
bool ActorGrantsValidArea(Actor a)
{
	return AreaTypes.Overlaps(a.TraitsImplementing<GivesBuildableArea>()
		.SelectMany(gba => gba.Types));
}
```

and give `BuildingInfo` and `GivesBuildableAreaInfo` a default value of `"Buildable"` in their type lists.  The extra complexity of supporting an empty set granting all types isn't worthwhile.

MustaphaTR commented:
Updated.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -40,6 +36,9 @@ public class BuildingInfo : ITraitInfo, IOccupySpaceInfo, IPlaceBuildingDecorati
 		public readonly int Adjacent = 2;
 
+		[Desc("Type of influences, defined by the `GivesBuildableArea` traits of the nearby actors, this can be build on.")]
+		public readonly HashSet<string> AreaTypes = new HashSet<string>() { "building" };
```
This is something I've always been against - magic default values for trait info's properties.

MustaphaTR replied:
Well, @pchote said me to give it a default.

penev92 replied:
I agree that it's not worth supporting an empty set. I don't agree with the magic default value though.
Leaving it empty and adding a `[FieldLoader.Require]` attribute should be enough. That will default the behaviour to "no buildings for you until you fix your missing YAML", which would be enforced by the attribute.

MustaphaTR replied:
I'm unsure about forcing it on Building:. What if it is unbuildable? Why would it need an area type?

GraionDilach replied:
A fallback value makes a lot more sense here than FieldLoader.Requiring every footprintable "building" (civilian structures, terrain doodads, trees, rocks, lamp posts) for GivesBuildableArea. When RA2 onwards the number of these assets are more than the actually buildable ones.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/GivesBuildableArea.cs:
```diff
@@ -0,0 +1,35 @@
+	public class GivesBuildableAreaInfo : ConditionalTraitInfo
+	{
+		[Desc("Accepted `AreaTypes` types. Leave empty to accept all types.")]
+		public readonly HashSet<string> Types = new HashSet<string>() { "building" };
```
Same with the default value.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/GivesBuildableArea.cs:
```diff
@@ -0,0 +1,35 @@
+	[Desc("Tag trait for actors with `AreaTypes`.")]
+	public class GivesBuildableAreaInfo : ConditionalTraitInfo
+	{
+		[Desc("Accepted `AreaTypes` types. Leave empty to accept all types.")]
```
`to accept` doesn't sound right, since this trait `gives`, it does not `accept`. ;)
Also at a glance it doesn't seem like leaving this empty will grant you any buildable area at all.

MustaphaTR replied:
It indeed doesn't. It should be complately removed.

penev92 reviewed (commented):
A few minor nits to clear up.

MustaphaTR commented:
About the default value thing, and also to remove hack/special case added with #13845, i think i can move Adjacent to own trait that something like `ChecksAdjacency:`, which requires `Building:`so instead of -1 special case and on unbuildable buildings we can just not use that trait. And also give it [FileLoader.Require] with no problem.

penev92 commented:
Yes, that sounds great, thanks :)

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/GivesBuildableArea.cs:
```diff
@@ -0,0 +1,35 @@
+	public class GivesBuildableAreaInfo : ConditionalTraitInfo
+	{
+		[Desc("Given area types to be checked by `Building` traits of actors while being placed.")]
+		public readonly HashSet<string> Types = new HashSet<string>() { "building" };
```
Also can you change this to `AreaTypes` to match the one on `Building`/`RequiresBuildableArea` and because upgrade rules can be a pain to write sometimes.

MustaphaTR commented:
Updated, but not done yet. I'm aware that i messed the rebase up, i'll fix that. This also needs a Upgrade Rule, and i only manually upgraded the RA mod, so others wouldn't work right now.

Also added a test case that removes `RequiresBuildableArea:` from Kennel.

MustaphaTR commented:
I unfortunately couldn't manage to write a sane upgrade rule for seperating Ajacent from Building. If anyone can write it please do. I'll just write one that adds AreaTypes: to GivesBuildableArea and fix other stuff manually for now.

MustaphaTR commented:
Fixed travis, Lonestar was missing AreaTypes: for GivesBuildableArea: on its custom rules (i didn't run upgrade rule for RA, so it was missed).

penev92 reviewed (commented):
I left a few minor comments.

penev92 commented:
Sorry, I know this comes a bit too late, but here's an upgrade rule based on yours:
```
if (engineVersion < 20171113)
{
	// AreaTypes support is added to GivesBuildableArea and it is required.
	var givesBuildableArea = node.Value.Nodes.FirstOrDefault(n => n.Key == "GivesBuildableArea");
	if (givesBuildableArea != null)
	{
		givesBuildableArea.Value.Nodes.Add(new MiniYamlNode("AreaTypes", "building"));
	}

	// RequiresBuildableArea trait is added and Building.Adjacent is moved there.
	var building = node.Value.Nodes.FirstOrDefault(n => n.Key == "Building");
	if (building != null)
	{
		var adjacent = building.Value.Nodes.FirstOrDefault(n => n.Key == "Adjacent");
		var areaTypes = new MiniYamlNode("AreaTypes", "building");
		var requiresBuildableArea = new MiniYamlNode("RequiresBuildableArea", "");

		requiresBuildableArea.Value.Nodes.Add(areaTypes);
		if (adjacent != null)
			requiresBuildableArea.Value.Nodes.Add(adjacent);

		node.Value.Nodes.Add(requiresBuildableArea);
		building.Value.Nodes.Remove(adjacent);
	}
}

```

MustaphaTR commented on OpenRA.Mods.Common/Traits/Buildings/RequiresBuildableArea.cs:
```diff
@@ -0,0 +1,34 @@
+		[Desc("Types of buildable are this actor requires.")]
+		public readonly HashSet<string> AreaTypes = new HashSet<string>();
+
+		[Desc("The range to the next building it can be constructed. Set it higher for walls.")]
```
It is not really "not tied to building" as there is `Requires<BuildingInfo>`, but will change.

MustaphaTR commented:
Updated.

penev92 reviewed (approved):
Code looks good to me now, and I'm keeping my fingers crossed for the YAML changes.

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f4f2fded36cc633caca0b624ff61e16169fe587b...ee0c5737d402d6530e44f3a32f6393e183900ba6)
